toolkit/tools/pkg/imagecustomizerlib/partitionutils.go

```diff
@@ -791,3 +808,153 @@ func getPartitionNum(partitionLoopDevice string) (int, error) {
 
 	return num, nil
 }
+
+// listBtrfsSubvolumes uses btrfs tools to discover actual subvolumes in a mounted BTRFS filesystem
+func listBtrfsSubvolumes(mountPoint string) ([]string, error) {
+	stdout, stderr, err := shell.Execute("btrfs", "subvolume", "list", "-a", mountPoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list BTRFS subvolumes: %w\nstderr: %s", err, stderr)
+	}
+
+	var subvolumes []string
+	lines := strings.Split(strings.TrimSpace(stdout), "\n")
+
+	for _, line := range lines {
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+
+		// Parse btrfs subvolume list output format:
+		// ID 256 gen 7 top level 5 path @
+		// ID 257 gen 7 top level 5 path @home
+		parts := strings.Fields(line)
+		if len(parts) >= 2 && parts[len(parts)-2] == "path" {
+			subvolPath := parts[len(parts)-1]
+			// Skip special subvolumes that start with <
+			if !strings.HasPrefix(subvolPath, "<") {
+				subvolumes = append(subvolumes, subvolPath)
+			}
+		}
+	}
+
+	logger.Log.Debugf("Found BTRFS subvolumes: %v", subvolumes)
+	return subvolumes, nil
+}
+
+// mountBtrfsSubvolume mounts a specific BTRFS subvolume
+func mountBtrfsSubvolume(devicePath, subvolume, mountPoint string) (*safemount.Mount, error) {
+	mountOptions := fmt.Sprintf("subvol=%s,ro", subvolume)
+	mount, err := safemount.NewMount(devicePath, mountPoint, "btrfs", unix.MS_RDONLY, mountOptions, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount BTRFS subvolume %s:\n%w", subvolume, err)
+	}
+	return mount, nil
+}
+
+// processBtrfsSubvolumes is a helper function that handles the common pattern of mounting BTRFS,
+// listing subvolumes, and iterating through them with a callback function
+func processBtrfsSubvolumes(diskPartition *diskutils.PartitionInfo, tmpDir string,
+	callback func(subvolume string, subvolMount *safemount.Mount) (bool, error),
+) (bool, error) {
+	// First mount the BTRFS filesystem to discover subvolumes
+	partitionMount, err := safemount.NewMount(diskPartition.Path, tmpDir, diskPartition.FileSystemType,
+		unix.MS_RDONLY, "", true)
+	if err != nil {
+		return false, fmt.Errorf("failed to mount partition (%s):\n%w", diskPartition.Path, err)
+	}
+	defer partitionMount.Close()
+
+	// List actual subvolumes using btrfs tools
+	subvolumes, err := listBtrfsSubvolumes(tmpDir)
+	if err != nil {
+		logger.Log.Debugf("Failed to list BTRFS subvolumes for partition %s:\n%v", diskPartition.Path, err)
```

Comment:
Don't ignore all errors.

Under what condition does `listBtrfsSubvolumes` fail but it is ok? Does `btrfs subvolume list` fail if there are no subvolumes? If so, then detect this case in `listBtrfsSubvolumes` and then return an empty list instead of an error.